### ISSUE TITLE
dispose() not called when FlutterViewController is popped

### ISSUE THREAD

**Issue by zhouyangyng** [flutter/flutter#42643]
I push a FlutterViewController from native vc on iOS app, and then pop the FlutterViewController. But the dispose() method is not called, so I can't destroy something and cause memory leaks.

Here is the flutter doctor log:

```
[✓] Flutter (Channel master, v1.9.6-pre.44, on Mac OS X 10.14.6
    18G95, locale zh-Hans-CN)
    • Flutter version 1.9.6-pre.44 at /Users/xmly/flutter
    • Framework revision d3b70c91bf (6 weeks ago), 2019-08-27
      19:08:35 -0700
    • Engine revision f93b9cbff1
    • Dart version 2.5.0


[✓] Android toolchain - develop for Android devices (Android SDK
    version 29.0.1)
    • Android SDK at /Users/xmly/Library/Android/sdk
    • Android NDK location not configured (optional; useful for
      native profiling support)
    • Platform android-29, build-tools 29.0.1
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_152-release-1343-b01)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 10.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 10.3, Build version 10G8
    • CocoaPods version 1.7.5

[✓] Android Studio (version 3.4)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 37.1.1
    • Dart plugin version 183.6270
    • Java version OpenJDK Runtime Environment (build
      1.8.0_152-release-1343-b01)

[✓] VS Code (version 1.38.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.4.1

[✓] Connected device (1 available)
    • iPhone SE   • 09C906C2-D2F0-4D59-B6BD-D929FA079E7D • ios •
      com.apple.CoreSimulator.SimRuntime.iOS-12-4 (simulator)

• No issues found!
```

**Comment by BondarenkoStas** [comment]
Similar to https://github.com/flutter/flutter/issues/28879.

**Comment by zhouyangyng** [comment]
> Similar to #28879.

But there is no solution in  #28879.

**Comment by talisk** [comment]
@zhouyangyng Could you post your FlutterEngine init code? What value did you pass in the `allowHeadlessExecution`?

**Comment by cbracken** [comment]
Can you take a look at @mehmetf's response to https://github.com/flutter/flutter/issues/28134 and see if that resolves your issue?

**Comment by zhouyangyng** [comment]
> @zhouyangyng Could you post your FlutterEngine init code? What value did you pass in the `allowHeadlessExecution`?

I'm sorry to reply you just now. Here is the init code:

```
FlutterViewController *flutterViewController = [[FlutterViewController alloc] init];
[self setChannelPlatform:flutterViewController];
[GeneratedPluginRegistrant registerWithRegistry:flutterViewController];
flutterViewController.splashScreenView = nil;
[[self currentNavigationController] pushViewController:flutterViewController animated:YES];
```

**Comment by zhouyangyng** [comment]
> Can you take a look at @mehmetf's response to #28134 and see if that resolves your issue?

I have read the issue and that is about dispose not called when app quit. However my problem is 
just when native viewcontroller pop not app quit.

**Comment by Hixie** [comment]
We could provide a hook that runs when the engine is about to be cleanly shut down (as opposed to when the OS kills the app), `Window.onEngineShutdown` or something, and then call `runApp(SizedBox.shrink())` by default in that callback... we'd need some way to make that not racy, either the framework synchronously doing the "cool down frame", or the engine immediately pushing one last frame...

**Comment by Hixie** [comment]
I'm not really convinced this should be a blocker. You could work around this by sending a message on the message channel saying "hey, clean up" without the engine itself doing it. For this reason, I'm removing the critical label.

**Comment by talisk** [comment]
@zhouyangyng The latest comment at [44256](https://github.com/flutter/flutter/issues/44256) may resolve your question. Make the FlutterEngine singleton, and pass to the init method of FlutterViewController. When the instance of FlutterViewController call dealloc, a timer starts. After a few seconds, call destroyContext.

So the FlutterViewController called dealloc, some code at the end can be executed.

**Comment by xster** [comment]
We also had a bunch of leaks on iOS that were since fixed. Could you try your test on the latest version of beta and 1.12.13?

**Comment by gaaclarke** [comment]
I believe this is fixed, a bit of cleanup has been added to the deletion of view controllers since filing this issue: https://github.com/flutter/engine/blob/5c9c6276c14c0d8a4c02ecc803768a66931111d5/shell/platform/darwin/ios/framework/Source/FlutterEngine.mm#L221

For example: https://github.com/flutter/engine/pull/13396

**Comment by xster** [comment]
I'm going to tentatively close this since we fixed a number of leak issues in 1.12 and there has been no new repros. Please feel free to open a new issue with reproductions steps if still present.
